Library/Homebrew/upgrade.rb

```diff
@@ -208,6 +208,7 @@ def install_formula(formula_installer, upgrade:)
       # possible for the existing build to interfere with the build we are about to
       # do! Seriously, it happens!
       kegs.each(&:unlink) if kegs.present?
+      return if formula_installer.dry_run?
```

Comment:
Dry run is already handled here so I don't think this is the right approach yet.